Creating a client with empty `redirectUris` or `webOrigins` causes ORA-01400 (NULL insert) on Oracle database

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

core

### Describe the bug

When using **Oracle** as the Keycloak database, creating a client with an empty string (`""`) in `redirectUris` or `webOrigins` causes an `ORA-01400: cannot insert NULL` constraint violation, resulting in a **409 Conflict** response with a misleading "Duplicate resource error" message.

This happens because **Oracle treats empty strings as NULL**, and the `REDIRECT_URIS.VALUE` and `WEB_ORIGINS.VALUE` columns have a `NOT NULL` constraint. Other databases (e.g. PostgreSQL) do not exhibit this behavior since they distinguish between empty strings and NULL.

### Version

26.6.3

### Regression

- [ ] The issue is a regression

### Expected behavior

Keycloak should either:
1. Validate and reject empty strings in `redirectUris` / `webOrigins` **before** persisting to the database (returning a proper 400 Bad Request with a meaningful error message), or
2. Filter out empty string entries before inserting into the database.

The behavior should be consistent across all supported databases.

### Actual behavior


- Keycloak attempts to insert an empty string into `REDIRECT_URIS.VALUE` or `WEB_ORIGINS.VALUE`.
- Oracle interprets the empty string as `NULL`, violating the `NOT NULL` constraint.
- Hibernate throws a `ConstraintViolationException`, which is then converted to a `ModelDuplicateException`.
- The API returns **HTTP 409 Conflict** with the message `"Duplicate resource error"`, which is misleading -- the actual problem is a NULL constraint violation, not a duplicate entry.
- The Keycloak admin console shows: *"Could not create client: 'conflict' — Duplicate resource error"*

### How to Reproduce?


**1. Environment setup**

Use the following `compose.yml` to start Oracle + Keycloak:

```yaml
services:
  oracle:
    image: container-registry.oracle.com/database/free:23.26.1.0
    environment:
      ORACLE_PWD: keycloak123
    ports:
      - "1521:1521"
    volumes:
      - oracle_data:/opt/oracle/oradata
      - ./init-oracle.sql:/opt/oracle/scripts/startup/init-oracle.sql
#   postgres:
#     image: postgres:16
#     environment:
#       POSTGRES_PASSWORD: keycloak123
#       POSTGRES_USER: keycloak
#       POSTGRES_DB: keycloak
#     ports:
#       - "5432:5432"
#     volumes:
#       - postgresql_data:/var/lib/postgresql/data

  keycloak:
    build:
      context: .
      dockerfile: Containerfile
    image: keycloak-oracle:latest
    command:
      - "start-dev"
    ports:
      - "8081:8080"
    environment:
      KC_DB: oracle
      KC_DB_URL: jdbc:oracle:thin:@//oracle:1521/FREEPDB1
      # KC_DB: postgres
      # KC_DB_URL: jdbc:postgresql://postgres:5432/keycloak
      KC_DB_USERNAME: keycloak
      KC_DB_PASSWORD: keycloak123
      KC_BOOTSTRAP_ADMIN_USERNAME: admin
      KC_BOOTSTRAP_ADMIN_PASSWORD: admin
      # KC_LOG_LEVEL: DEBUG

volumes:
  oracle_data:
  postgresql_data:

```

`init-oracle.sql`:

```sql
ALTER SESSION SET CONTAINER = FREEPDB1;
ALTER SESSION SET "_ORACLE_SCRIPT"=true;

CREATE USER keycloak IDENTIFIED BY "keycloak123"
  DEFAULT TABLESPACE users TEMPORARY TABLESPACE temp
  QUOTA UNLIMITED ON users;

GRANT CREATE SESSION, CREATE TABLE, CREATE SEQUENCE,
      CREATE VIEW, CREATE PROCEDURE, CREATE TRIGGER,
      CREATE TYPE, UNLIMITED TABLESPACE TO keycloak;
GRANT SELECT ON SYS.DBA_RECYCLEBIN TO keycloak;

```

`Containerfile`:

```dockerfile
FROM quay.io/curl/curl:latest AS driver
RUN curl -L -o /tmp/ojdbc17.jar \
  https://repo1.maven.org/maven2/com/oracle/database/jdbc/ojdbc17/23.26.2.0.0/ojdbc17-23.26.2.0.0.jar

FROM quay.io/keycloak/keycloak:26.6
ENV KC_DB=oracle
COPY --from=driver /tmp/ojdbc17.jar /opt/keycloak/providers/ojdbc17.jar
RUN /opt/keycloak/bin/kc.sh build
```

**2. Reproduce via Admin CLI (`kcadm.sh`)**

```bash
# Authenticate
kcadm.sh config credentials --server http://localhost:8080 --realm master --user admin --password admin

# This triggers the error — empty redirectUris and webOrigins
kcadm.sh create clients -r master \
  -s clientId=client-error \
  -s 'redirectUris=[""]' \
  -s 'webOrigins=[""]' \
  -s publicClient=false \
  -s directAccessGrantsEnabled=true -i
```

**3. Reproduce via Admin Console**

Create a new client in the Keycloak admin console and leave the **Redirect URIs** or **Web Origins** fields with an empty entry (just add an entry and leave it blank).

**4. Reproduce via kcadm.sh**

Create a new client using kcadm.sh and leave the **Redirect URIs** or **Web Origins** fields with an empty entry.

```bash
$HOME/bin/kcadm.sh config credentials --server http://localhost:8080 --realm master --user admin --password admin
$HOME/bin/kcadm.sh create clients -r master -s clientId=client-error -s 'redirectUris=[""]' -s 'webOrigins=[""]' -s 'publicClient=false' -s 'directAccessGrantsEnabled=true' -i
```


### Anything else?

### Error logs

**Scenario 1 — Empty `redirectUris` only:**

```
Caused by: org.hibernate.exception.ConstraintViolationException:
  could not execute batch [ORA-01400: cannot insert NULL into ("KEYCLOAK"."REDIRECT_URIS"."VALUE")]
  [insert into REDIRECT_URIS (CLIENT_ID,VALUE) values (?,?)]
```

**Scenario 2 — Empty `webOrigins` only:**

```
Caused by: org.hibernate.exception.ConstraintViolationException:
  could not execute batch [ORA-01400: cannot insert NULL into ("KEYCLOAK"."WEB_ORIGINS"."VALUE")]
  [insert into WEB_ORIGINS (CLIENT_ID,VALUE) values (?,?)]
```

Both are ultimately caused by:

```
java.sql.BatchUpdateException: ORA-01400: cannot insert NULL into ("KEYCLOAK"."REDIRECT_URIS"."VALUE")
```

The `ConstraintViolationException` is incorrectly converted to `ModelDuplicateException` ("Duplicate resource error"), making the error message misleading.

### Root cause analysis

1. Oracle treats **empty strings (`''`) as NULL** — this is a well-known Oracle behavior and differs from PostgreSQL, MySQL, and other databases.
2. The `REDIRECT_URIS.VALUE` and `WEB_ORIGINS.VALUE` columns are `NOT NULL`.
3. Keycloak does not validate or sanitize empty string entries in `redirectUris` / `webOrigins` before persisting.
4. The JPA exception converter in `EntityManagerProxy.convert()` maps the `ConstraintViolationException` to `ModelDuplicateException`, producing a misleading 409 Conflict instead of a proper validation error.

### Affected version

- Keycloak / RHBK: **26.6.3**
- Oracle Database: **23ai Free (23.26.1.0)**
- JDBC Driver: **ojdbc17 23.26.2.0.0**

### Additional notes

- This issue does **not** occur with PostgreSQL, as PostgreSQL correctly distinguishes between empty strings and NULL values.
- The same behavior likely affects any other column where Keycloak stores user-provided strings that could be empty when using Oracle as the database backend.

--- a/server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java
+++ b/server-spi-private/src/main/java/org/keycloak/models/utils/RepresentationToModel.java
@@ -19,6 +19,7 @@ package org.keycloak.models.utils;
 
 import java.io.IOException;
 import java.util.AbstractMap;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -62,7 +63,6 @@ import org.keycloak.client.clienttype.ClientTypeException;
 import org.keycloak.client.clienttype.ClientTypeManager;
 import org.keycloak.common.Profile;
 import org.keycloak.common.Profile.Feature;
-import org.keycloak.common.util.CollectionUtil;
 import org.keycloak.common.util.MultivaluedHashMap;
 import org.keycloak.common.util.ObjectUtil;
 import org.keycloak.common.util.UriUtils;
@@ -541,8 +541,8 @@ public class RepresentationToModel {
             // Client Secret
             add(updatePropertyAction(client::setSecret, () -> determineNewSecret(client, rep)));
             // Redirect uris / Web origins
-            add(updatePropertyAction(client::setRedirectUris, () -> CollectionUtil.collectionToSet(rep.getRedirectUris()), client::getRedirectUris));
-            add(updatePropertyAction(client::setWebOrigins, () -> CollectionUtil.collectionToSet(rep.getWebOrigins()), () -> defaultWebOrigins(client)));
+            add(updatePropertyAction(client::setRedirectUris, () -> filterEmptyValues(rep.getRedirectUris()), client::getRedirectUris));
+            add(updatePropertyAction(client::setWebOrigins, () -> filterEmptyValues(rep.getWebOrigins()), () -> defaultWebOrigins(client)));
         }};
 
         // Extended client attributes
@@ -617,6 +617,13 @@ public class RepresentationToModel {
                 .stream()
                 .filter(uri -> uri.startsWith("http"))
                 .map(UriUtils::getOrigin)
+                .collect(Collectors.toSet());
+    }
+
+    private static Set<String> filterEmptyValues(Collection<String> collection) {
+        if (collection == null) return null;
+        return collection.stream()
+                .filter(s -> s != null && !s.isBlank())
                 .collect(Collectors.toSet());
     }
 

--- a/tests/base/src/test/java/org/keycloak/tests/admin/ClientTest.java
+++ b/tests/base/src/test/java/org/keycloak/tests/admin/ClientTest.java
@@ -82,6 +82,7 @@ import org.keycloak.testframework.realm.RealmConfig;
 import org.keycloak.testframework.realm.RoleBuilder;
 import org.keycloak.testframework.realm.UserBuilder;
 import org.keycloak.testframework.util.ApiUtil;
+import org.keycloak.tests.suites.DatabaseTest;
 import org.keycloak.tests.utils.Assert;
 import org.keycloak.tests.utils.admin.AdminApiUtil;
 import org.keycloak.tests.utils.admin.AdminEventPaths;
@@ -1234,5 +1235,50 @@ public class ClientTest {
         rep.getAttributes().put(OIDCConfigAttributes.CLIENT_SESSION_MAX_LIFESPAN, "950");
         createOrUpdateClientExpectingValidationErrors(rep, false,
                 "Client session max lifespan cannot exceed realm SSO session max lifespan and RememberMe Max span.");
+    }
+
+    @Test
+    @DatabaseTest
+    public void createClientWithEmptyRedirectUrisAndWebOrigins() {
+        ClientRepresentation rep = new ClientRepresentation();
+        rep.setClientId("empty-uris-client");
+        rep.setEnabled(true);
+        rep.setRedirectUris(Arrays.asList("", "https://example.com", "  "));
+        rep.setWebOrigins(Arrays.asList("", "https://example.com", "  "));
+
+        Response response = managedRealm.admin().clients().create(rep);
+        assertEquals(201, response.getStatus());
+        String id = ApiUtil.getCreatedId(response);
+        managedRealm.cleanup().add(r -> r.clients().get(id).remove());
+        adminEvents.poll();
+
+        ClientRepresentation stored = managedRealm.admin().clients().get(id).toRepresentation();
+        assertEquals(Set.of("https://example.com"), new HashSet<>(stored.getRedirectUris()));
+        assertEquals(Set.of("https://example.com"), new HashSet<>(stored.getWebOrigins()));
+    }
+
+    @Test
+    @DatabaseTest
+    public void updateClientWithEmptyRedirectUrisAndWebOrigins() {
+        ClientRepresentation rep = new ClientRepresentation();
+        rep.setClientId("update-empty-uris-client");
+        rep.setEnabled(true);
+        rep.setRedirectUris(Arrays.asList("https://example.com"));
+        rep.setWebOrigins(Arrays.asList("https://example.com"));
+
+        Response response = managedRealm.admin().clients().create(rep);
+        assertEquals(201, response.getStatus());
+        String id = ApiUtil.getCreatedId(response);
+        managedRealm.cleanup().add(r -> r.clients().get(id).remove());
+        adminEvents.poll();
+
+        rep.setRedirectUris(Arrays.asList("", "https://updated.com", "  "));
+        rep.setWebOrigins(Arrays.asList("", "https://updated.com", "  "));
+        managedRealm.admin().clients().get(id).update(rep);
+        adminEvents.poll();
+
+        ClientRepresentation stored = managedRealm.admin().clients().get(id).toRepresentation();
+        assertEquals(Set.of("https://updated.com"), new HashSet<>(stored.getRedirectUris()));
+        assertEquals(Set.of("https://updated.com"), new HashSet<>(stored.getWebOrigins()));
     }
 }

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/SecureRedirectUrisEnforcerExecutorTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/SecureRedirectUrisEnforcerExecutorTest.java
@@ -17,9 +17,12 @@
 package org.keycloak.testsuite.client.policies;
 
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 import org.keycloak.OAuthErrorException;
 import org.keycloak.client.registration.ClientRegistrationException;
@@ -101,10 +104,10 @@ public class SecureRedirectUrisEnforcerExecutorTest extends AbstractClientPolici
         try {
             updateClientByAdmin(cId, (ClientRepresentation clientRep) -> {
                 clientRep.setAttributes(new HashMap<>());
-                clientRep.setRedirectUris(List.of("")); // nomally, vacant redirect uri is not allowed.
+                clientRep.setRedirectUris(List.of("")); // empty redirect uris are filtered out before persistence.
             });
             ClientRepresentation cRep = getClientByAdmin(cId);
-            assertEquals(new HashSet<>(List.of("")), new HashSet<>(cRep.getRedirectUris()));
+            assertEquals(Collections.emptySet(), new HashSet<>(cRep.getRedirectUris()));
         } catch (ClientPolicyException cpe) {
             fail();
         }
@@ -695,7 +698,10 @@ public class SecureRedirectUrisEnforcerExecutorTest extends AbstractClientPolici
                 clientRep.setRedirectUris(redirectUrisList);
             });
             ClientRepresentation cRep = getClientByAdmin(alphaCid);
-            assertEquals(new HashSet<>(redirectUrisList), new HashSet<>(cRep.getRedirectUris()));
+            Set<String> expectedUris = redirectUrisList.stream()
+                    .filter(s -> s != null && !s.isBlank())
+                    .collect(Collectors.toSet());
+            assertEquals(expectedUris, new HashSet<>(cRep.getRedirectUris()));
         } catch (ClientPolicyException cpe) {
             fail();
         }

